service/src/main/java/uk/nhs/adaptors/gpc/consumer/filters/FindAPatientDocsGatewayFilterFactory.java

```diff
@@ -0,0 +1,111 @@
+import org.springframework.cloud.gateway.filter.GatewayFilter;
+import org.springframework.cloud.gateway.filter.GatewayFilterChain;
+import org.springframework.cloud.gateway.filter.OrderedGatewayFilter;
+import org.springframework.cloud.gateway.filter.factory.AbstractGatewayFilterFactory;
+import org.springframework.core.io.buffer.DataBuffer;
+import org.springframework.core.io.buffer.DataBufferUtils;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.server.reactive.ServerHttpResponse;
+import org.springframework.http.server.reactive.ServerHttpResponseDecorator;
+import org.springframework.stereotype.Component;
+import org.springframework.web.server.ServerWebExchange;
+
+import lombok.Getter;
+import lombok.Setter;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import reactor.core.publisher.Mono;
+import uk.nhs.adaptors.gpc.consumer.utils.FindAPatientDocsUtil;
+
+@Component
+@Slf4j
+public class FindAPatientDocsGatewayFilterFactory extends AbstractGatewayFilterFactory<FindAPatientDocsGatewayFilterFactory.Config> {
+    private static final List<String> LOGGABLE_HEADER_KEYS = List.of("Ssp-From", "Ssp-To");
+    private static final String LOG_TEMPLATE = "Gateway filter log: %s URL: %s";
+    private static final String HEADERS_PREFIX = "Headers: { ";
+    private static final String HEADERS_SUFFIX = "}";
+    private static final String COLON = ": ";
+    private static final int PRIORITY = -2;
+
+    public FindAPatientDocsGatewayFilterFactory() {
+        super(FindAPatientDocsGatewayFilterFactory.Config.class);
+    }
+
+    @Override
+    public GatewayFilter apply(Config config) {
+        return new OrderedGatewayFilter((exchange, chain) -> prepareGatewayFilterMono(exchange, chain, config), PRIORITY);
+    }
+
+    private String prepareHeaderLog(HttpHeaders httpHeaders) {
+        StringBuilder headersLogBuilder = new StringBuilder(HEADERS_PREFIX);
+        LOGGABLE_HEADER_KEYS.forEach(key -> {
+            if (httpHeaders.containsKey(key)) {
+                headersLogBuilder.append(key)
+                    .append(COLON)
+                    .append(httpHeaders.get(key))
+                    .append(StringUtils.SPACE);
+            }
+        });
+        headersLogBuilder.append(HEADERS_SUFFIX);
+
+        return headersLogBuilder.toString();
+    }
+
+    @SneakyThrows
+    private Mono<Void> handleError(ServerHttpResponse response, DataBuffer dataBuffer, Config config) {
+        if (response != null && dataBuffer != null) {
+            if (isErrorResponseCode(response)) {
+                LOGGER.error("An error with status occurred: " + response.getStatusCode());
+                LOGGER.error(StandardCharsets.UTF_8.decode(dataBuffer.asByteBuffer()).toString());
+            } else {
+                var decompressedResponseString = FindAPatientDocsUtil.unzipInputStreamToString(dataBuffer.asInputStream());
```

Comment:
How are you sure the content is compressed? You could check the 'Content-Encoding' heading.

The better way to handle this would be get spring cloud gateway to give you a string - it will handle encoding correctly.

Reply:
You recieve an error header miss match if it is not encoded

As discussed implementing .modifyResponseBody